activerecord/lib/active_record/transactions.rb

```diff
@@ -338,7 +338,7 @@ def before_committed! # :nodoc:
     # Ensure that it is not called if the object was never persisted (failed create),
     # but call it after the commit of a destroyed object.
     def committed!(should_run_callbacks: true) #:nodoc:
```

Comment:
Since this only appears to happen with `after_commit` and not `after_update_commit` does it make sense to change how `after_commit` defines callbacks instead of changing `committed!`?

Reply:
Good question.

You are correct that `after_update_commit` is not prone to this buggy behavior, but I believe that is incidental, not intentional. I even added a spec to this PR to cover that behavior because even though it works, it's not explicitly accounted for. 

Did you have a specific suggestion for how to approach this from the `after_commit` side rather than the `committed!` side?

Reply:
`after_update_commit` and related all look like this: https://github.com/rails/rails/blob/268dd6966ddc6f66bf09fb96046bd435e0515b6e/activerecord/lib/active_record/transactions.rb#L244-L247
`after_commit` looks like this: https://github.com/rails/rails/blob/268dd6966ddc6f66bf09fb96046bd435e0515b6e/activerecord/lib/active_record/transactions.rb#L232-L235
Could you do something like this?
```ruby
def after_commit(*args, &block)
  options   = args.extract_options!
  on_option = options.extract!(:on)
  set_options_for_callbacks!(args, on_option)
  set_callback(:commit, :after, *args, **options, &block)
end
```

Reply:
Sorry, just not following how that fixes anything.

Reply:
I was confused. I thought `after_commit :cb, on: :update` was the problem but its with contextless `after_commit` callbacks only, correct?  My bad. I wonder if defaulting to all validation contexts in `set_options_for_callbacks!` will work:
```ruby
def set_options_for_callbacks!(args, enforced_options = {})
  options = args.extract_options!.merge!(enforced_options)
  args << options
  options
  
  fire_on = Array(options.fetch(:on, %i(create update destroy)))
  assert_valid_transaction_action(fire_on)
  options[:if] = Array(options[:if])
  options[:if].unshift(-> { transaction_include_any_action?(fire_on) })
end
```